posts/fall2019/js-chats-3/index.md

```diff
@@ -0,0 +1,1285 @@
+code into a package and publish it online for other 
+people to use it!
+(Check out rehooks.com. It is a collection of hooks
+that has been published as NPM packages.)
+
+### What's Wrong with the Old `class` Way?
+
+In a `class` component, logic can be distributed across 
+multiple lifecycle methods.
+
+![demonstration of spread out logic in class component](images/spread-out-logic.png)
+
+It is hard to abstraction the logic out to a reusable
+standalone component/function.
+
+However, the logic is compact with hooks.
+
+![demonstration of compact logic in function component](images/compact-logic.png)
+
+If the code that belongs to the same object is close
+together, it is easy to abstract them into their own 
+custom hooks or functions. The new pattern enables
+you to play with logic as LEGO blocks and insert them
+into any place you want. 
+
+You can read about the motivation behind the new hooks API
+from the React team: https://reactjs.org/docs/hooks-intro.html#motivation
+
+## Final Challenge
+
+I am going to build a counter that increments itself every X second.
+
+```jsx
+function SelfIncrementingCounter() {
+  const [count, setCount] = useState(0);
+  useEffect(() => {
+    setInterval(() => {
+      setCount(count + 1);
+    }, 200);
+  });
+
+  return <h1>Count: {count}</h1>;
+}
+```
+
+> Note: `setInterval` schedule a callback to execute 
+> for some interval. In this case, `setCount` is called 
+> every 200 ms.
+
+But this implementation does not work. Can you explain what is the
+wrong behavior? 
+
+<details>
+
+<summary>Click to see answer</summary>
+
+![glitching behavior of counter](images/final-challenge-wrong-1.gif)
+
+`useEffect` is called on every render, since there is no
+dependency array specified. 
+Each time `useEffect` is executed, it start a repeated scheduled task.
+The number of repeated scheduled tasks grows exponentially. 
+There are races between the different tasks leading to a glitch.
+
+</details>
+
+The previous attempt failed. I try again, passing an empty
+dependency array such that the `useEffect` callback is only
+executed once after the first render (like `componentDidMount`).
+
+```jsx
+function SelfIncrementingCounter() {
+  const [count, setCount] = useState(0);
+  useEffect(() => {
+    setInterval(() => {
+      setCount(count + 1);
+    }, 200);
+  }, []); // added empty dependency array here
+
+  return <h1>Count: {count}</h1>;
+}
+```
+
+Does this work? (Hint: it does not) Why?
+
+This one might be a bit hard. 
+Here is some hint:
+- There can be multiple copies of `count`. Why?
+- There can be multiple versions of the callback function
+  passed to `useEffect`, which one does `useEffect` use?
+- In the executed version of callback function, which 
+  copy of `count` is the `count` referring to?
+
+Now we know that it works, can we implement a version
+that works? Hint: the setter `setXXX` function for state
+can also takes a function as a parameter.
```

Comment:
Could another way be:

```js
function SelfIncrementingCounter() {
  const [count, setCount] = useState(0);
  useEffect(() => {
    setTimeout(() => {  // here we use a single timeout
      setCount(count + 1);
    }, 200);
  }, [count]);  // here we declare a dependency on count
  return <h1>Count: {count}</h1>;
}
```

I feel like the `setState(oldCount => oldCount + 1)` pattern is generally preferred and there might be edge cases where this fails to work, but just wanted to hear your thoughts.

Reply:
I kept it this way since part of final challenge is to let people figure out the function pattern

Reply:
Oh, no I'm not asking you to change it. I'm wondering if what I wrote would work, as I'm not familiar with hooks, is all I meant.

Reply:
Oh, yeah. The correct answer is to use the function pattern with `setInterval`, since it does not depend on the previous function frame's value of `count`. 

Using `setTimeout` also works. But a better solution is to return a clean up function that clears the last timeout handle. Since a component can unmount for some reason and it will trigger an update for an unmounted component. (The same applies for `setInterval` as well)